### PR TITLE
Add pthread support

### PR DESCRIPTION
**Pull request by coolyjg**
- Add `pthread` related libc function
  - [x] create, exit, join
  - [x] pthread_self
  - [ ] mutex_init, mutex_lock, mutex_unlock
- Add `pthread` tests and `pipe` tests

### CODE REVIEW

**Comment by chyyuu** [comment]
Good work!

File: ulib/libax/src/cbindings/thread.rs [hunk 1]
```diff
@@ -8,3 +14,227 @@ pub unsafe extern "C" fn ax_getpid() -> c_int {
...
+    }
+
+    fn get_void(&self) -> *mut c_void {
+        self.inner
+    }
+}
+
+/// `pthread_setcanceltype` implementation
+#[no_mangle]
+pub unsafe extern "C" fn ax_pthread_setcanceltype(new: c_int, old: *mut c_int) -> c_int {
```
Comment (equation314) [inline ulib/libax/src/cbindings/thread.rs]: Not used in c_libax?

File: ulib/libax/src/cbindings/thread.rs [hunk 2]
```diff
@@ -8,3 +14,227 @@ pub unsafe extern "C" fn ax_getpid() -> c_int {
...
     })
 }
+
+pub const AX_THREAD_LIMIT: usize = 1024;
+const THREAD_NONE: Option<Mutex<SendSyncWrapper>> = None;
+
+/// Mapping between `pid` to `pthread_t`
+///
+/// Since each thread only access their own `pthread_t`, this can be lockless.
+static mut THREAD_TABLE: [Option<Mutex<SendSyncWrapper>>; AX_THREAD_LIMIT] =
```
Comment (equation314) [inline ulib/libax/src/cbindings/thread.rs]: Use `RwLock<BTreeMap<u64, ...>>`

File: ulib/libax/src/cbindings/thread.rs [hunk 3]
```diff
@@ -8,3 +14,227 @@ pub unsafe extern "C" fn ax_getpid() -> c_int {
...
+pub unsafe extern "C" fn ax_pthread_setcanceltype(new: c_int, old: *mut c_int) -> c_int {
+    debug!("ax_pthread_setcanceltype <= new: {}, old: {}", new, *old);
+    if new as u32 > 1u32 {
+        return LinuxError::EINVAL as c_int;
+    }
+    let tid = current().id().as_u64();
+    if tid == 2 && THREAD_TABLE[tid as usize].is_none() {
+        add_main_thread();
+    }
+    let mut ptr = Box::from_raw(
```
Comment (equation314) [inline ulib/libax/src/cbindings/thread.rs]: `ptr` will be dropped when there function returns, dangerous!

File: ulib/libax/src/cbindings/thread.rs (hunk 2 above)
Comment (equation314) [reply ulib/libax/src/cbindings/thread.rs]: And implement a function like `pthread_self`

**Comment by chyyuu** [comment]
@coolyjg Good work! @equation314 OK?

File: ulib/libax/src/cbindings/thread.rs (hunk 1 above)
Comment (coolyjg) [reply ulib/libax/src/cbindings/thread.rs]: Solved and update header

File: ulib/libax/src/cbindings/thread.rs (hunk 2 above)
Comment (coolyjg) [reply ulib/libax/src/cbindings/thread.rs]: Change into `RwLock<BTreeMap<u64, SendSyncWrapper>>`, and use `lazy_static`.
Add `pthread_self` implementation.

File: ulib/libax/src/cbindings/thread.rs (hunk 3 above)
Comment (coolyjg) [reply ulib/libax/src/cbindings/thread.rs]: Fix this by `core::mem::forget`

File: ulib/libax/build.rs
Comment (equation314) [inline ulib/libax/build.rs:1]: Add `config.function.no_return = Some("__attribute__((noreturn))".into());` in `gen_rust_to_c_bindings`.

File: ulib/c_libax/src/pthread.c [hunk 4]
```diff
@@ -2,27 +2,54 @@
 #include <pthread.h>
 #include <unistd.h>
 
-#if defined(AX_CONFIG_MULTITASK)
+#if defined(AX_CONFIG_MULTITASK) && defined(AX_CONFIG_ALLOC)
 _Noreturn void pthread_exit(void *result)
 {
-    unimplemented();
```
Comment (equation314) [inline ulib/c_libax/src/pthread.c:8]: Call `ax_pthread_exit(result)` without `return`.

Comment (coolyjg) [reply ulib/c_libax/src/pthread.c:8]: Solved

File: apps/c/pthread/parallel/main.c [hunk 5]
```diff
@@ -1,61 +1,74 @@
...
 
-void main()
+int main()
 {
-    pthread_t t1;
-    pthread_t t2;
 
-    pthread_create(&t1, NULL, ThreadFunc1, NULL);
-    pthread_create(&t2, NULL, ThreadFunc2, NULL);
+    for (int i = 0; i < NUM_DATA; i++) array[i] = i;
```
Comment (equation314) [inline apps/c/pthread/parallel/main.c]: `array[i] = rand()`

File: apps/c/pthread/parallel/main.c [hunk 6]
```diff
@@ -1,61 +1,74 @@
...
+        tasks[i] = t;
+    }
+
+    uint64_t *partial_sum;
+    uint64_t actual = 0;
+    for (int i = 0; i < NUM_TASKS; i++) {
+        pthread_join(tasks[i], (void *)(&partial_sum));
+        actual += *partial_sum;
+    }
+
```
Comment (equation314) [inline apps/c/pthread/parallel/main.c:66]: output the sum

File: ulib/c_libax/include/pthread.h [hunk 7]
```diff
@@ -26,6 +26,8 @@ typedef struct {
         volatile int __vi[sizeof(long) == 8 ? 10 : 6];
         volatile void *volatile __p[sizeof(long) == 8 ? 5 : 6];
     } __u;
+    // Inner ArceOS mutex number
+    int mutex_no;
```
Comment (equation314) [inline ulib/c_libax/include/pthread.h]: Recover this change

File: apps/c/pthread/parallel/main.c [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+int main()
+{
+
+    for (int i = 0; i < NUM_DATA; i++) array[i] = i;
+
+    uint64_t expect = 0;
+    for (int i = 0; i < NUM_DATA; i++) expect += my_sqrt(array[i]);
+
+    int thread_part[NUM_TASKS];
```
Comment (equation314) [inline apps/c/pthread/parallel/main.c:51]: This array is unnecessary?

**Review by equation314 — APPROVED** [review]

**Pull request merged**